### PR TITLE
Abstract message encoding functions

### PR DESCRIPTION
**Pull request by gr8bit**
# Work in Progress - do not merge

### What is the purpose of this pull request?
Abstracted the JSON message encoding into a single interface (MessageEncoder) so it can be replaced with custom encoders more easily. See #91 

### Checklist

- [ ] I've added tests for this change
- [ ] I've added a Changelog entry
- [ ] I've updated documentation

### CODE REVIEW

File: comm/json_encoder.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+func (jp jsonEncoder) MarshalTransmissions(transmissions []string, message *common.Message) ([][]byte, error) {
+	var transmissionBytes [][]byte
+	for _, transmission := range transmissions {
+		transmissionBytes = append(transmissionBytes, []byte(transmission))
+	}
+
+	return transmissionBytes, nil
+}
+
+func (jp jsonEncoder) MarshalAuthenticateTransmissions(transmissions []string) ([][]byte, error) {
```
Comment (palkan) [inline comm/json_encoder.go:40]: Why do we need to treat these transmissions separately?

File: Makefile [hunk 2]
```diff
@@ -14,7 +14,7 @@ ifdef VERSION
 else
 	COMMIT := $(shell sh -c 'git log --pretty=format:"%h" -n 1 ')
 	VERSION := $(shell sh -c 'git tag -l --sort=-version:refname "v*" | head -n1')
-	LD_FLAGS="-s -w -X github.com/anycable/anycable-go/utils.sha=$(COMMIT) -X github.com/anycable/anycable-go/utils.version="
+	LD_FLAGS="-s -w -X github.com/anycable/anycable-go/utils.sha=$(COMMIT) -X github.com/anycable/anycable-go/utils.version=$(VERSION)"
```
Comment (palkan) [inline Makefile:17]: Good catch!

Would be great to have it an a separate PR to merge faster)

File: common/common.go [hunk 3]
```diff
@@ -138,6 +138,26 @@ type RemoteDisconnectMessage struct {
...
+	Reconnect bool   `json:"reconnect"`
+}
+
+// Reply represents outgoing client message
+type Reply struct {
+	Type       string      `json:"type,omitempty"`
+	Identifier string      `json:"identifier"`
+	Message    interface{} `json:"message"`
+}
+
```
Comment (palkan) [inline common/common.go:160]: Also looks like a good enhancement on its own, so, feel free to extract into a separate PR.

File: Makefile (hunk 2 above)
Comment (gr8bit) [reply Makefile:17]: Added #115

File: common/common.go (hunk 3 above)
Comment (gr8bit) [reply common/common.go:160]: Added in #116

File: comm/json_encoder.go (hunk 1 above)
Comment (gr8bit) [reply comm/json_encoder.go:35]: Only to be able to provide the encoder with some context about the transmissions' origin. These are possible after authenticate reply (and there's no other context) or in a call reply (in which case there's a common.Message as context). That's why I split them into two separate API calls, one being called on authenticate reply, the other on call reply (with the message).
However, I just removed some duplicate code in the json encoder implementation.
